posix/signal.c

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <openenclave/enclave.h>
+
+#include <openenclave/corelibc/errno.h>
+#include <openenclave/corelibc/signal.h>
+#include <openenclave/corelibc/stdlib.h>
+#include <openenclave/corelibc/unistd.h>
+#include <openenclave/internal/calls.h>
+#include <openenclave/internal/posix/lock.h>
+#include <openenclave/internal/posix/raise.h>
+#include <openenclave/internal/print.h>
+#include <openenclave/internal/thread.h>
+#include <openenclave/internal/trace.h>
+#include "posix_t.h"
+
+// Poll uses much of the infrastructure from epoll.
+
+static void _handle_ignore(int signum);
+
+static struct oe_sigaction _actions[__OE_NSIG] = {{{0}}};
+
+static oe_sighandler_t _default_actions[__OE_NSIG] = {
+    _handle_ignore, _handle_ignore, _handle_ignore, _handle_ignore,
+    _handle_ignore, _handle_ignore, _handle_ignore, _handle_ignore,
+    _handle_ignore, _handle_ignore, _handle_ignore, _handle_ignore,
+    _handle_ignore, _handle_ignore, _handle_ignore, _handle_ignore,
+    _handle_ignore, _handle_ignore, _handle_ignore, _handle_ignore,
+    _handle_ignore, _handle_ignore, _handle_ignore, _handle_ignore,
+    _handle_ignore, _handle_ignore, _handle_ignore, _handle_ignore,
+    _handle_ignore, _handle_ignore, _handle_ignore, _handle_ignore};
+
+static oe_spinlock_t _lock = OE_SPINLOCK_INITIALIZER;
+
+static void _handle_ignore(int signum)
+{
+    (void)signum;
+}
+
+static void _handle_error(int signum)
+{
+    (void)signum;
+}
+
+int oe_kill(oe_pid_t pid, int signum)
+{
+    int retval = -1;
+    oe_errno = 0;
+
+    if (oe_posix_kill_ocall(&retval, (int)pid, signum) != OE_OK)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    retval = 0;
+
+done:
+    return retval;
+}
+
+int oe_sigaction(
+    int signum,
+    const struct oe_sigaction* act,
+    struct oe_sigaction* oldact)
+{
+    int retval = -1;
+    bool locked = false;
+
+    oe_register_posix_ecall_function_table();
+
+    if (signum >= __OE_NSIG)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    oe_conditional_lock(&_lock, &locked);
+
+    if (oldact)
+    {
+        *oldact = _actions[signum];
+    }
+
+    if (act)
+    {
+        _actions[signum] = *act;
+    }
+
+    retval = 0;
+done:
+
+    oe_conditional_unlock(&_lock, &locked);
+
+    return retval;
+}
+
+oe_sighandler_t oe_signal(int signum, oe_sighandler_t handler)
+{
+    oe_sighandler_t retval = OE_SIG_ERR;
+    bool locked = false;
+
+    oe_register_posix_ecall_function_table();
+
+    if (signum >= __OE_NSIG)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    oe_conditional_lock(&_lock, &locked);
+    _actions[signum].__oe_sigaction_handler.oe_sa_handler = handler;
+
+done:
+    oe_conditional_unlock(&_lock, &locked);
+
+    return retval;
+}
+
+int oe_posix_signal_notify_ecall(int signum)
```

Comment:
How is this supposed to work? Signals trigger an AEX that for an existing AEP we already register that dispatches to our vectored exception handler (see enclave/core/sgx/exception.c). The host doesn't get to register a second one to invoke this ecall. The signal handling needs to be dispatched from there and reconciled with the handlers registered through the `oe_add_vectored_exception_handler`/`oe_remove_vectored_exception_handler` methods.

This seems architecturally broken ... nothing actually refers to signals in this changeset, so can we take this out entirely until we have an integrated design with the existing exception handling implementation?

Reply:
Removed singal feature from this PR.